core/src/main/java/org/infinispan/persistence/manager/PersistenceManagerImpl.java

```diff
@@ -1384,25 +1388,42 @@ private long getMaxEntries() {
       return Long.MAX_VALUE;
    }
 
-   private void preloadKey(AdvancedCache<Object, Object> cache, MarshallableEntry me) {
-      final Transaction transaction = suspendIfNeeded();
-      boolean success = false;
-      try {
+   private Single<?> preloadKey(AdvancedCache<Object, Object> cache, MarshallableEntry me) {
+      // CallInterceptor will preserve the timestamps if the metadata is an InternalMetadataImpl instance
+      InternalMetadataImpl metadata = new InternalMetadataImpl(me.getMetadata(), me.created(), me.lastUsed());
+      CompletionStage<Object> stage;
+      if (configuration.transaction().transactionMode().isTransactional() && transactionManager != null) {
+         final Transaction transaction = suspendIfNeeded();
+         CompletionStage<Transaction> putStage;
          try {
             beginIfNeeded();
-            // CallInterceptor will preserve the timestamps if the metadata is an InternalMetadataImpl instance
-            InternalMetadataImpl metadata = new InternalMetadataImpl(me.getMetadata(), me.created(), me.lastUsed());
-            cache.put(me.getKey(), me.getValue(), metadata);
-            success = true;
+            putStage = cache.putAsync(me.getKey(), me.getValue(), metadata)
+               .thenApply(ignore -> {
+                  try {
+                     return transactionManager.suspend();
+                  } catch (SystemException e) {
+                     throw new PersistenceException("Unable to preload!", e);
+                  }
+               });
          } catch (Exception e) {
             throw new PersistenceException("Unable to preload!", e);
-         } finally {
-            commitIfNeeded(success);
          }
-      } finally {
-         //commitIfNeeded can throw an exception, so we need a try { } finally { }
-         resumeIfNeeded(transaction);
+         stage = (CompletionStage) putStage.whenCompleteAsync((pendingTransaction, t) -> {
+            try {
+               transactionManager.resume(pendingTransaction);
+               commitIfNeeded(t == null);
+            } catch (InvalidTransactionException | SystemException e) {
+               throw new PersistenceException("Unable to preload!", e);
+            } finally {
+               //commitIfNeeded can throw an exception, so we need a try { } finally { }
+               resumeIfNeeded(transaction);
+            }
+         }, blockingExecutor);
+      } else {
+         stage = cache.putAsync(me.getKey(), me.getValue(), metadata);
       }
+      return RxJavaInterop.completionStageToMaybe(stage)
+            .toSingle(me);
```

Comment:
The return value isn't used, so `completionStageToSingle()` should work -- or there should be a comment explaining why the double conversion is needed.

Reply:
It is used sadly because stage can have a null value and we have to have a non null one. Same reason why I can't use `CompletionStage<?>` above.